contracts/BC_fusion/StakePool.sol

```diff
@@ -0,0 +1,262 @@
+        uint256 bnbAmount = _burnAndSync(delegator, shares);
+
+        uint256 _gasLimit = IStakeHub(STAKE_HUB_ADDR).transferGasLimit();
+        (bool success,) = STAKE_HUB_ADDR.call{ gas: _gasLimit, value: bnbAmount }("");
+        require(success, "TRANSFER_FAILED");
+
+        emit Unbonded(delegator, shares, bnbAmount);
+        return bnbAmount;
+    }
+
+    function claim(address payable delegator, uint256 number) external onlyStakeHub nonReentrant returns (uint256) {
+        // number == 0 means claim all
+        // number should not exceed the length of the queue
+        require(_unbondRequestsQueue[delegator].length() != 0, "NO_UNBOND_REQUEST");
+        number = (number == 0 || number > _unbondRequestsQueue[delegator].length())
+            ? _unbondRequestsQueue[delegator].length()
+            : number;
+
+        uint256 _totalShares;
+        uint256 _totalBnbAmount;
+        while (number != 0) {
+            bytes32 hash = _unbondRequestsQueue[delegator].front();
+            UnbondRequest memory request = _unbondRequests[hash];
+            if (block.timestamp < request.unlockTime) {
+                break;
+            }
+
+            _totalShares += request.shares;
+            _totalBnbAmount += request.bnbAmount;
+
+            // remove from the queue
+            _unbondRequestsQueue[delegator].popFront();
+            delete _unbondRequests[hash];
+
+            --number;
+        }
+        require(_totalShares != 0, "NO_CLAIMABLE_UNBOND_REQUEST");
+        _lockedShares[delegator] -= _totalShares;
+
+        uint256 _gasLimit = IStakeHub(STAKE_HUB_ADDR).transferGasLimit();
+        (bool success,) = delegator.call{ gas: _gasLimit, value: _totalBnbAmount }("");
+        require(success, "CLAIM_FAILED");
+
+        emit UnbondClaimed(delegator, _totalShares, _totalBnbAmount);
+        return _totalBnbAmount;
+    }
+
+    function distributeReward(uint64 commissionRate) external payable onlyStakeHub {
+        uint256 bnbAmount = msg.value;
+        uint256 _commission = (bnbAmount * uint256(commissionRate)) / COMMISSION_RATE_BASE;
+        uint256 _reward = bnbAmount - _commission;
+        totalPooledBNB += _reward;
+
+        // mint commission to the validator
+        _mintAndSync(validator, _commission);
+
+        emit RewardReceived(_reward, _commission);
+    }
+
+    function slash(uint256 slashBnbAmount) external onlyStakeHub returns (uint256) {
+        uint256 selfDelegation = balanceOf(validator);
+        uint256 slashShares = getSharesByPooledBNB(slashBnbAmount);
+
+        uint256 _remainingSlashBnbAmount;
+        if (slashShares <= selfDelegation) {
+            _burnAndSync(validator, slashShares);
+        } else {
+            uint256 selfDelegationBNB = _burnAndSync(validator, selfDelegation);
+            unchecked {
+                // slashShares > selfDelegation
+                _remainingSlashBnbAmount = slashBnbAmount - selfDelegationBNB;
```

Comment:
remove "unchecked" just for a good sleep

Reply:
let us simplify it:

```
        uint256 selfDelegation = balanceOf(validator);
        uint256 slashShares = getSharesByPooledBNB(slashBnbAmount);
        slashShares = slashShares>selfDelegation? selfDelegation:slashShares
        realSlashBnbAmount = _burnAndSync(validator, slashShares);
        emit Slashed(realSlashBnbAmount);
```

Reply:
Cool